Document multi-pool IPAM mode

tklauser opened the pull request:
The first two commits are cleanups following the deprecation of the `cluster-pool-v2beta` IPAM mode (ref. #25767).

The third commit adds a tutorial showing the use of the multi-pool IPAM mode and the `CiliumPodIPPool` CRD and also document the concepts and known limitations of the `multi-pool` IPAM mode.

For #25470

tklauser committed:
docs/bgp-control-plan: remove mention of `cluster-pool-v2beta` IPAM mode

The `cluster-pool-v2beta` IPAM mode was deprecated in commit
d18cf579816a ("docs: Deprecate `cluster-pool-v2beta`") and related docs
were removed. Remove it from suggested IPAM modes in the BGP control
plane docs as well.

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
docs/ipam: remove deprecated Cluster Scope v2 from IPAM mode table

The `cluster-pool-v2beta` IPAM mode was deprecated in commit
d18cf579816a ("docs: Deprecate `cluster-pool-v2beta`") and related docs
were removed. Remove it from the IPAM mode table as well.

Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

learnitall commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -0,0 +1,153 @@
+  3. If neither the pod nor the namespace have a ``ipam.cilium.io/ip-pool`` annotation,
+     the pod's IP will be allocated from the pool named ``default``.
+
+The annotation is only considered when the pod is created. Changing the ``ip-pool``
```
```suggestion
The annotation is only considered when a pod is created. Changing the ``ip-pool``
```

learnitall commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -0,0 +1,153 @@
+``ipam-multi-pool-pre-allocation`` map are assumed to have a ``preAllocIPs`` of zero,
+i.e. no IPs are pre-allocated for that pool.
+
+The formula Cilium agent uses to compute the number of needed IPs each each pool is:
```
```suggestion
The formula Cilium agent uses to compute the number of needed IPs from each pool is:
```

learnitall commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -0,0 +1,153 @@
+
+  neededIPs = roundUp(inUseIPs + pendingIPs + preAllocIPs, preAllocIPs)
+
+Where ``inUseIPs`` is the number of IPs that are currently actively in use,
```
I think we should choose "currently" or "actively", but not both.

tklauser replied:
Changed to "currently in use".

learnitall commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -0,0 +1,153 @@
+pool. This behavior can be controlled using the
+``ipam-multi-pool-pre-allocation`` flag.  It contains a key-value map of the
+form ``<pool-name>=<preAllocIPs>`` where ``preAllocIPs`` specifies how many IPs
+(both IPv4 and IPv6) are to be pre-allocated to the local node.
```
Can we change the wording here to make it more clear how IPv4 and IPv6 IPs are picked from a pool? I can users interpreting one of the following given the language used here (given `preAllocIPs=X`)

* Mix of IPv4 and IPv6 IPs to total `X`.
* `X` IPv4 and `X` IPv6 IPs.

tklauser replied:
I've reworked this section to make it (hopefully) more clear that `X` IPs are allocated for each IP address family in case the pool contains CIDRs for both IPv4 and IPv6.

learnitall replied:
Look great, thank you!

learnitall commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -0,0 +1,153 @@
+Allocation Parameters
+---------------------
+
+Cilium agent can be configured to pre-allocate a minimum number of IPs from each
```
The word minimum sticks out here. Is there a situation where the Cilium agent will pre-allocate more than the number of configured IPs from each pool? This language is also used in other places in this section, but further along the paragraph we have `specifies how many IPs (both IPv4 and IPv6 are to be pre-allocated to the local node.`, which implies that the number of pre-allocated IPs is a set value and not just a minimum.

tklauser replied:
I've dropped the term "minimum" from this section because I think it is not entirely correct anyway. For the purpose of explaining the pre-allocation algorithm I don't think it adds value. The exact mechanism is already documented in the formula below.

learnitall replied:
Looks great, thank you! There is still one usage of `at least` in this context on line 117. Can we remove that one as well?

tklauser replied:
Dropped this as well, thanks.

learnitall commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -0,0 +1,153 @@
+***********
+
+Multi-Pool IPAM is a preview feature. The following limitations apply to Cilium running in
+``multi-pool`` IPAM mode:
```
```suggestion
Multi-Pool IPAM mode:
```

This will match the way the feature is referred to at the beginning of the document.

learnitall commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -0,0 +1,153 @@
+
+.. warning::
+   - Tunnel mode is not supported. Multi-Pool IPAM may only be used in direct routing mode.
+   - Transparent encryption with IPSec is not supported.
```
Can we add information about Wireguard here as well?

tklauser replied:
This is part of the "Limitations" section. AFAIK there are no known limitations with Wireguard, so I'm not sure what we want to add here re. Wireguard?

learnitall replied:
Ah I see. I think it would still be useful if we add a note making it clear that Wireguard is fully supported, even if IPSec isn't. Something like: 

```Transparent encryption is only supported with Wireguard and cannot be used with IPSec.```

tklauser replied:
Changed to "Transparent encryption is only supported with Wireguard and cannot be used with IPSec."

learnitall commented on Documentation/network/kubernetes/ipam-multi-pool.rst:
```diff
@@ -0,0 +1,194 @@
+
+       [...]
+
+#. Validate that the ``CiliumPodIPPool`` resource for the ``default`` pool got created with the
```
```suggestion
#. Validate that the ``CiliumPodIPPool`` resource for the ``default`` pool was created with the
```

learnitall commented on Documentation/network/kubernetes/ipam-multi-pool.rst:
```diff
@@ -0,0 +1,194 @@
+               - containerPort: 80
+       EOF
+
+#. Validate that the pods got assigned IPv4 addresses from different CIDRs as specified in the pool
```
```suggestion
#. Validate that the pods were assigned IPv4 addresses from different CIDRs as specified in the pool
```

learnitall commented on Documentation/network/kubernetes/ipam-multi-pool.rst:
```diff
@@ -0,0 +1,194 @@
+       $ kubectl create namespace cilium-test
+       $ kubectl annotate namespace cilium-test ipam.cilium.io/ip-pool=mars
+
+   All pods created in namespace ``cilium-test`` get assigned IPv4 addresses from the ``mars`` pool.
```
```suggestion
   All new pods created in the namespace ``cilium-test`` will be assigned IPv4 addresses from the ``mars`` pool.
```

learnitall commented on Documentation/network/kubernetes/ipam-multi-pool.rst:
```diff
@@ -0,0 +1,194 @@
+       [...]
+       ✅ All 42 tests (295 actions) successful, 13 tests skipped, 0 scenarios skipped.
+
+#. Verify that the connectivity test pods got assigned IPv4 addresses from the 10.20.0.0/16 CIDR
```
```suggestion
#. Verify that the connectivity test pods were assigned IPv4 addresses from the 10.20.0.0/16 CIDR
```

learnitall commented on Documentation/network/kubernetes/ipam-multi-pool.rst:
```diff
@@ -0,0 +1,194 @@
+   * ``--set kubeProxyReplacement=strict``
+   * ``--set bpf.masquerade=true``
+
+   For more details on why each of these these options are needed, please refer to
```
```suggestion
   For more details on why each of these options are needed, please refer to
```

learnitall commented on Documentation/network/kubernetes/ipam-multi-pool.rst:
```diff
@@ -0,0 +1,194 @@
+   * ``--set endpointRoutes.enabled=true``
+   * ``--set-string extraConfig.enable-local-node-route=false``
+   * ``--set kubeProxyReplacement=strict``
+   * ``--set bpf.masquerade=true``
```
I think `--set tunnel=disabled` is missing here.

tklauser replied:
Nice catch, added.

learnitall reviewed (changes requested):
Looks great! Most of these are small things like nits, typos, and wording, but I think there are a couple of things missing as well that we should add.

This is my first review as part of the docs-structure team, so let's also get another review here to double-check.

qmonnet commented on Documentation/network/kubernetes/ipam-multi-pool.rst:
```diff
@@ -0,0 +1,194 @@
+
+
+
+
```
```suggestion
```

qmonnet commented on Documentation/network/concepts/ipam/multi-pool.rst:
```diff
@@ -0,0 +1,153 @@
+
+.. _ipam_crd_multi_pool:
+
+Multi-Pool (Beta)
+#################
```
Given that it's beta, let's add the standard disclaimer?

```suggestion
Multi-Pool (Beta)
#################

.. include:: ../../../beta.rst
```

(Same for `ipam-multi-pool.rst`)

qmonnet reviewed (approved):
Looks good as far as I'm concerned, although I didn't check how you addressed Ryan's feedback - I'm deferring to Ryan for that :).

tklauser committed:
Document multi-pool IPAM mode

Add a tutorial showing the use of the multi-pool IPAM mode and the
CiliumPodIPPool CRD. Also document the concepts and limitations of this
IPAM mode.

Co-authored-by: Sebastian Wicki <sebastian@cilium.io>
Signed-off-by: Sebastian Wicki <sebastian@cilium.io>
Signed-off-by: Tobias Klauser <tobias@cilium.io>

learnitall reviewed (approved):
Looks good to go!
